### PR TITLE
adds helper function to check for an arango collection index

### LINKED ISSUE #1722: Handling cases where a collection's index is missing or renamed
**Describe the bug**
`isOccurrence` collection used to have an index named `byPkgIDArtIDJust` which was renamed to `byPkgIDArtIDOriginJust` in September 2023 under [this](https://github.com/guacsec/guac/commit/88dfb7ea1c9b4e2f023c082f152ef22138e832b0#diff-1aba4420a9017b9d486c0fab38a08b9eea2e9c22e2402039d7726f26b1ccf544L560) commit. The change also includes addition of the field `origin` to the index.

However, since guac [creates indices](https://github.com/guacsec/guac/blob/main/pkg/assembler/backends/arangodb/backend.go#L141-L151) only if the `guac` graph doesn't already exist, the whole process of creating new/renamed indices is skipped when we already have a database and are not starting from scratch.

This issue is not just about creating new collection indices but also about managing the old ones. Should we delete all the existing indices that are not in our current map? Should we edit them? Or are there better options?

**To Reproduce**

1. Clone the repo and switch to the [80eab75](https://github.com/guacsec/guac/tree/80eab75cc78ff433a5cc3cf6b4e8d873391f8bd5) commit state
2. Run `make container` to create the `local-organic-guac` image
3. Switch to the `main` branch
4. Update the `‎guac-graphql‎` container in [arango.yaml](https://github.com/guacsec/guac/blob/main/container_files/arango.yaml) file to refer to the image created in step#2
5. In Arango DB UI check that the index `byPkgIDArtIDJust` is created
6. Ingest a provenance and verify the corresponding entry shows up in the `isOccurrence` collection
7. Update the `‎guac-graphql‎` container image to `local-organic-guac` image built out of the `main` branch
8. Check the indices of the `isOccurrence` collection

**Expected behavior**
`isOccurrence` collection should have only `byPkgIDArtIDOriginJust` index

**Ingested document(s)**
Any sample slsa provenance

Can you share the documents that are used to reproduce the ingestion errors or showcase the data quality issues.
No

### PR DESCRIPTION
**Pull request by semmet95**
# Description of the PR

Currently when `guac` graph exists in arango backend, creation of collection indexes is skipped, this leads to missing/redundant indexes in the database after upgrades. This PR adds the logic to re-index collections if the `guac` graph exists. The PR includes following changes:
- Check existing collection indexes against the expected collection indexes. Delete the unexpected ones and create the missing expected indexes.
- Check if vertex collections of an edge collection exist. Create the edge collection if any of them don't.

Fixes #1722

# PR Checklist

- [x] All commits have [a Developer Certificate of Origin (DCO)](https://wiki.linuxfoundation.org/dco) -- they are generated using `-s` flag to `git commit`.
- [ ] All new changes are covered by tests
- [ ] If GraphQL schema is changed, `make generate` has been run
- [ ] If OpenAPI spec is changed, `make generate` has been run
- [ ] If `collectsub` protobuf has been changed, `make proto` has been run
- [ ] All CI checks are passing (tests and formatting)
- [x] All dependent PRs have already been merged

### CODE REVIEW

File: pkg/assembler/backends/arangodb/backend.go [hunk 1]
```diff
@@ -475,6 +273,22 @@ func createIndexPerCollection(ctx context.Context, db driver.Database, collectio
 	return nil
 }
 
+func deleteIndexPerCollection(ctx context.Context, db driver.Database, collection string, indexName string) error {
+	databaseCollection, err := db.Collection(ctx, collection)
+	if err != nil {
+		return err
```
Comment (pxp928) [inline pkg/assembler/backends/arangodb/backend.go:279]: can you add more descriptive errors via `fmt.Errorf`

File: pkg/assembler/backends/arangodb/backend.go [hunk 2]
```diff
@@ -475,6 +273,22 @@ func createIndexPerCollection(ctx context.Context, db driver.Database, collectio
...
+	}
+
+	idx, err := databaseCollection.Index(ctx, indexName)
+	if err != nil {
+		return err
```
Comment (pxp928) [inline pkg/assembler/backends/arangodb/backend.go:284]: descriptive error

File: pkg/assembler/backends/arangodb/backend.go [hunk 3]
```diff
@@ -475,6 +273,22 @@ func createIndexPerCollection(ctx context.Context, db driver.Database, collectio
...
+	}
+
+	err = idx.Remove(ctx)
```
Comment (pxp928) [inline pkg/assembler/backends/arangodb/backend.go:287]: either do return `idx.Remove(ctx)` or return a descriptive error (if error) or return nil

File: pkg/assembler/backends/arangodb/backend.go [hunk 4]
```diff
@@ -652,14 +466,225 @@ func createGraph(ctx context.Context, db driver.Database, graphName string, edge
...
+		// Note:: even though the edge collection exists, its corresponding vertex collections might not
+		var vertexCollections []string
+		vertexCollections = append(vertexCollections, edgeDefinition.From...)
+		vertexCollections = append(vertexCollections, edgeDefinition.To...)
+
+		// create the edge collection if any vertex collection is missing
+		for _, vertexCollection := range vertexCollections {
+			exists, err := graph.VertexCollectionExists(ctx, vertexCollection)
+			if err != nil {
+				return err
```
Comment (pxp928) [inline pkg/assembler/backends/arangodb/backend.go:488]: descriptive error

File: pkg/assembler/backends/arangodb/backend.go [hunk 5]
```diff
@@ -652,14 +466,225 @@ func createGraph(ctx context.Context, db driver.Database, graphName string, edge
...
+		}
+	}
+	return nil
+}
+
+func getCollectionIndexMap() map[string][]index {
+	collectionIndexMap := make(map[string][]index)
+
+	collectionIndexMap[artifactsStr] = []index{
+		*(initIndex("byDigest", []string{"digest"}, true)),
```
Comment (pxp928) [inline pkg/assembler/backends/arangodb/backend.go:506]: is there a reason why you are taking a reference only to de-reference it for the map? Should change to:

```
func initIndex(name string, fields []string, unique bool) index {
	return index{
		name: name,
		fields: fields,
		unique: unique,
	}
}
```

File: pkg/assembler/backends/arangodb/backend.go [hunk 6]
```diff
@@ -652,14 +466,225 @@ func createGraph(ctx context.Context, db driver.Database, graphName string, edge
...
+
+		for _, index := range indexes {
+			indexExpected[index.name] = true
+		}
+
+		// get all the existing indexes for the collection
+		fmt.Printf("fetching collection: %s from the arango db\n", collectionName)
+		collection, err := arangoDb.Collection(ctx, collectionName)
+		if err != nil {
+			return err
```
Comment (pxp928) [inline pkg/assembler/backends/arangodb/backend.go:660]: descriptive error

File: pkg/assembler/backends/arangodb/backend.go [hunk 7]
```diff
@@ -652,14 +466,225 @@ func createGraph(ctx context.Context, db driver.Database, graphName string, edge
...
+		}
+
+		fmt.Printf("fetching all the indexes for the collection: %s\n", collectionName)
+		existingIndexes, err := collection.Indexes(ctx)
+		if err != nil {
+			return err
```
Comment (pxp928) [inline pkg/assembler/backends/arangodb/backend.go:666]: descriptive error

File: pkg/assembler/backends/arangodb/backend.go [hunk 8]
```diff
@@ -652,14 +466,225 @@ func createGraph(ctx context.Context, db driver.Database, graphName string, edge
...
+		}
+
+		fmt.Printf("re-indexing collection: %s\n", collectionName)
```
Comment (pxp928) [inline pkg/assembler/backends/arangodb/backend.go:669]: use logger

File: pkg/assembler/backends/arangodb/backend.go [hunk 9]
```diff
@@ -652,14 +466,225 @@ func createGraph(ctx context.Context, db driver.Database, graphName string, edge
...
+		for _, existingIndex := range existingIndexes {
+			if !indexExpected[existingIndex.UserName()] && existingIndex.Type() != "primary" {
+				// delete the outdated index
+				fmt.Printf("deleting unexpected index: %s for collection: %s\n", existingIndex.UserName(), collectionName)
```
Comment (pxp928) [inline pkg/assembler/backends/arangodb/backend.go:673]: logger

File: pkg/assembler/backends/arangodb/backend.go [hunk 10]
```diff
@@ -652,14 +466,225 @@ func createGraph(ctx context.Context, db driver.Database, graphName string, edge
...
+				deleteIndexPerCollection(ctx, arangoDb, collectionName, existingIndex.UserName())
+			} else {
+				indexExists[existingIndex.UserName()] = true
+			}
+		}
+
+		for _, index := range indexes {
+			// create the index if doesn't already exist
+			if !indexExists[index.name] {
+				fmt.Printf("creating index: %s for collection: %s\n", index.name, collectionName)
```
Comment (pxp928) [inline pkg/assembler/backends/arangodb/backend.go:683]: logger

**Review by pxp928 — COMMENTED** [review]
@semmet95 Added some comments. Thanks for the PR!

**Comment by semmet95** [comment]
@pxp928 Implemented all the suggestions

File: pkg/assembler/backends/arangodb/backend.go [hunk 11]
```diff
@@ -157,231 +176,11 @@ func getBackend(ctx context.Context, args backends.BackendArgs) (backends.Backen
...
-			return nil, fmt.Errorf("failed to generate guackey index for srcNames: %w", err)
-		}
-
-		// GUAC key indices for vulnerabilities
-		if err := createIndexPerCollection(ctx, db, vulnerabilitiesStr, []string{"guacKey"}, false, "byVulnGuacKey"); err != nil {
-			return nil, fmt.Errorf("failed to generate index for vulnerabilities: %w", err)
+		fullCollectionIndexMap := getCollectionIndexMap()
+		for collectionName, indexes := range fullCollectionIndexMap {
+			for _, index := range indexes {
+				createIndexPerCollection(ctx, db, collectionName, index.fields, index.unique, index.name)
```
Comment (mihaimaruseac) [inline pkg/assembler/backends/arangodb/backend.go:182]: Should we check error here?

**Review by mihaimaruseac — COMMENTED** [review]
Overall this looks good to me, but I'm wondering whether we should check errors in all places where we call a function that can error

File: pkg/assembler/backends/arangodb/backend.go (hunk 11 above)
Comment (pxp928) [reply pkg/assembler/backends/arangodb/backend.go:182]: yes, @semmet95 please fix all the lint issues.

Comment (semmet95) [reply pkg/assembler/backends/arangodb/backend.go]: Hi @mihaimaruseac 
you are correct, we should be checking the returned errors.
I have updated all such function calls to handle the returned error.

**Review by pxp928 — APPROVED** [review]
LGTM! Thanks @semmet95!

**Review by mihaimaruseac — APPROVED** [review]
Thank you!

**Pull request merged**